mps/mps.py

```diff
@@ -0,0 +1,574 @@
+      for site in range(self.center_position + 1, n2):
+        left_env = net.contract_between(left_env, nodes[site])
+        left_env = net.contract_between(left_env, conj_nodes[site])
+        if site + 1 in sites:
+          left_env.reorder_edges([edges[site], conj_edges[site]])
+          left_envs[site + 1] = left_env.tensor
+    return left_envs
+
+  def right_envs(self, sites: List[int]) -> Dict:
+    """
+    Compute right reduced density matrices for site `sites.
+    Args:
+      sites (list of int): A list of sites of the MPS.
+    Returns:
+      dict maping int to Tensors: The right-reduced density matrices 
+        at each  site in `sites`.
+
+    """
+
+    n1 = min(sites)
+    sites = np.array(sites)
+    if not np.all(np.array(sites) < len(self)):
+      raise ValueError('all elements of `sites` have to be < N = {}'.format(
+          len(self)))
+    if not np.all(np.array(sites) >= -1):
+      raise ValueError('all elements of `sites` have to be >= -1')
+
+    right_sites = sites[sites >= self.center_position]
+    right_envs = {}
+    for site in right_sites:
+      right_envs[site] = self.backend.eye(N=self.nodes[site].shape[2])
+
+    if n1 < self.center_position:
+      net = tensornetwork.TensorNetwork(
+          backend=self.backend.name, dtype=self.dtype)
+      nodes = {}
+      conj_nodes = {}
+      for site in reversed(range(n1 + 1, self.center_position + 1)):
+        nodes[site] = net.add_node(self.nodes[site].tensor)
+        conj_nodes[site] = net.add_node(
+            self.backend.conj(self.nodes[site].tensor))
+      nodes[self.center_position][2] ^ conj_nodes[self.center_position][2]
+      nodes[self.center_position][1] ^ conj_nodes[self.center_position][1]
+
+      for site in reversed(range(n1 + 1, self.center_position)):
+        nodes[site][2] ^ nodes[site + 1][0]
+        conj_nodes[site][2] ^ conj_nodes[site + 1][0]
+        nodes[site][1] ^ conj_nodes[site][1]
+
+      edges = {site: node[0] for site, node in nodes.items()}
+      conj_edges = {site: node[0] for site, node in conj_nodes.items()}
+
+      right_env = net.contract_between(nodes[self.center_position],
+                                       conj_nodes[self.center_position])
+      if self.center_position - 1 in sites:
+        right_env.reorder_edges(
+            [edges[self.center_position], conj_edges[self.center_position]])
+        right_envs[self.center_position - 1] = right_env.tensor
+      for site in reversed(range(n1 + 1, self.center_position)):
+        right_env = net.contract_between(right_env, nodes[site])
+        right_env = net.contract_between(right_env, conj_nodes[site])
+        if site - 1 in sites:
+          right_env.reorder_edges([edges[site], conj_edges[site]])
+          right_envs[site - 1] = right_env.tensor
+
+    return right_envs
+
+  def __len__(self):
+    return len(self.nodes)
+
+  def transfer_operator(self, site: int, direction: Union[Text, int],
```

Comment:
just thinking ahead, for infinite MPS the concept of transfer matrix makes sense as an object (although for efficiency, one might not want to construct it). For the sake of. a common API ,perhaps rename this to `apply_transfer_matrix`?

Reply:
renamed